lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -1978,6 +1978,57 @@ class DecomposeAtenFullOp : public OpRewritePattern<AtenFullOp> {
 };
 } // namespace
 
+namespace {
+// Decompose `aten.linear` op into `aten.matmul` and `aten.add` ops.
+class DecomposeAtenLinearOp : public OpRewritePattern<AtenLinearOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenLinearOp op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value input = op.input();
+    Value weight = op.weight();
+    Value bias = op.bias();
+
+    BaseTensorType inputType = input.getType().cast<BaseTensorType>();
+    if (!inputType.hasSizes() || inputType.getSizes().size() < 2)
+      return rewriter.notifyMatchFailure(
+          op, "expected input to be rank 2 or greater");
+
+    BaseTensorType weightType = weight.getType().cast<BaseTensorType>();
+    // weight must be a rank 2 matrix.
+    if (!weightType.hasSizes() || weightType.getSizes().size() != 2)
+      return rewriter.notifyMatchFailure(op, "expected weight to be a rank 2");
+
+    ArrayRef<int64_t> wShape = weightType.getSizes();
```

Comment:
This does not work with dynamic shapes.

Reply:
Oh wait, yeah it does. Sorry. Usually getSizes() is not good to use for dynamic shapes but in this case it does work.

Reply:
nit: I would recommend doing `llvm::to_vector(llvm::reverse(weightType.getSizes())` for this.